File: cmd/netspeed/output/output.go
// Package output provides terminal output formatting with ASCII spinners and colors.
package output

import (
	"encoding/json"
	"fmt"
	"os"
	"sort"
	"strings"
	"sync"
	"time"

	"github.com/yellowman/netspeed/cmd/netspeed/client"
)

// ANSI color codes
const (
	ColorReset  = "\033[0m"
	ColorRed    = "\033[31m"
	ColorGreen  = "\033[32m"
	ColorYellow = "\033[33m"
	ColorBlue   = "\033[34m"
	ColorCyan   = "\033[36m"
	ColorBold   = "\033[1m"
	ColorDim    = "\033[2m"
)

// Config holds output configuration.
type Config struct {
	JSON        bool
	CSV         bool
	Quiet       bool
	Verbose     bool
	Color       bool
	Interactive bool
}

// Output handles formatted terminal output.
type Output struct {
	cfg     Config
	spinner *Spinner
	mu      sync.Mutex
}

// New creates a new Output instance.
func New(cfg Config) *Output {
	return &Output{cfg: cfg}
}

// IsInteractive returns true if running in an interactive terminal.
func IsInteractive() bool {
	if os.Getenv("CI") != "" {
		return false
	}
	if os.Getenv("TERM") == "dumb" {
		return false
	}
	fi, err := os.Stdout.Stat()
	if err != nil {
		return false
	}
	return fi.Mode()&os.ModeCharDevice != 0
}

// color wraps text in ANSI color codes if colors are enabled.
func (o *Output) color(code, text string) string {
	if !o.cfg.Color {
		return text
	}
	return code + text + ColorReset
}

// Header prints the test header.
func (o *Output) Header(serverURL, version string) {
	if o.cfg.JSON || o.cfg.CSV || o.cfg.Quiet {
		return
	}

	fmt.Printf("%s %s\n", o.color(ColorBold, "netspeed"), displayVersion(version))
	fmt.Printf("Server: %s\n", o.color(ColorCyan, serverURL))
	fmt.Println(strings.Repeat("─", 48))
}

func displayVersion(version string) string {
	version = strings.TrimSpace(version)
	if version == "" {
		return "dev"
	}
	if version == "dev" || strings.HasPrefix(version, "v") {
		return version
	}
	return "v" + version
}

// Progress updates the progress display.
func (o *Output) Progress(phase string, current, total int, value float64) {
	if o.cfg.JSON || o.cfg.CSV || o.cfg.Quiet {
		return
	}

	o.mu.Lock()
	defer o.mu.Unlock()

	if !o.cfg.Interactive {
		// Non-interactive: just print progress
		if o.cfg.Verbose {
			fmt.Printf("%s: %d/%d (%.1f)\n", phase, current, total, value)
		}
		return
	}

	// Interactive: show progress bar
	percent := float64(current) / float64(total)
	barWidth := 20
	filled := int(percent * float64(barWidth))
	empty := barWidth - filled

	bar := strings.Repeat("=", filled)
	if filled < barWidth {
		bar += ">"
		empty--
	}
	bar += strings.Repeat(" ", max(0, empty))

	var valueStr string
	switch phase {
	case "download", "upload":
		valueStr = fmt.Sprintf("%.1f Mbps", value)
	case "latency":
		valueStr = fmt.Sprintf("%.1f ms", value)
	default:
		valueStr = fmt.Sprintf("%.1f", value)
	}

	label := strings.Title(phase)
	fmt.Printf("\r%-12s [%s] %3.0f%% %s\033[K", label+":", bar, percent*100, valueStr)
}

// ClearProgress clears the progress line.
func (o *Output) ClearProgress() {
	if o.cfg.Interactive {
		fmt.Print("\r" + strings.Repeat(" ", 60) + "\r")
	}
}

// Error prints an error message.
func (o *Output) Error(msg string) {
	if o.cfg.JSON {
		_ = json.NewEncoder(os.Stderr).Encode(map[string]string{"error": msg})
		return
	}
	fmt.Fprintf(os.Stderr, "%s %s\n", o.color(ColorRed, "Error:"), msg)
}

// Results prints the final results.
func (o *Output) Results(r *client.Results) {
	o.ClearProgress()

	fmt.Println()
	fmt.Println(strings.Repeat("─", 48))
	fmt.Printf("%s\n", o.color(ColorBold, "                    RESULTS"))
	fmt.Println(strings.Repeat("─", 48))

	// Download
	dlStr := fmt.Sprintf("%.1f Mbps", r.Summary.DownloadMbps)
	fmt.Printf("  Download:     %s\n", o.color(ColorCyan, dlStr))

	// Upload
	ulStr := fmt.Sprintf("%.1f Mbps", r.Summary.UploadMbps)
	fmt.Printf("  Upload:       %s\n", o.color(ColorCyan, ulStr))

	// Latency
	latStr := fmt.Sprintf("%.1f ms (jitter: %.1f ms)", r.Summary.LatencyUnloadedMs, r.Summary.JitterMs)
	fmt.Printf("  Latency:      %s\n", o.color(ColorBlue, latStr))

	// Packet loss. The compatibility summary is round-trip transaction loss;
	// Phase 2 also exposes forward probes and reverse acknowledgements.
	if r.PacketLoss != nil && r.PacketLoss.Unavailable {
		fmt.Printf("  Packet Loss:  %s\n", o.color(ColorYellow, "N/A ("+r.PacketLoss.Reason+")"))
	} else if r.PacketLoss != nil {
		plStr := fmt.Sprintf("%.2f%% transaction (%d/%d)",
			r.PacketLoss.TransactionLossPercent, r.PacketLoss.Received, r.PacketLoss.Sent)
		fmt.Printf("  Packet Loss:  %s\n", plStr)
		fmt.Printf("    Forward:    %s\n", formatDirectionalLoss(
			r.PacketLoss.ForwardLossPercent,
			r.PacketLoss.ForwardReceived,
			r.PacketLoss.ForwardSent,
		))
		fmt.Printf("    Reverse ACK:%s\n", formatDirectionalLoss(
			r.PacketLoss.ReverseAcknowledgementLossPercent,
			r.PacketLoss.AcknowledgementsReceived,
			r.PacketLoss.AcknowledgementsSent,
		))
	} else {
		fmt.Printf("  Packet Loss:  %s\n", o.color(ColorYellow, "N/A (skipped)"))
	}

	confidence := fmt.Sprintf("%s (%d/100)", strings.ToUpper(r.TestConfidence.Overall), r.TestConfidence.OverallScore)
	fmt.Printf("  Confidence:   %s\n", o.confidenceColor(r.TestConfidence.Overall, confidence))

	fmt.Println(strings.Repeat("─", 48))

	// Network Quality
	fmt.Printf("  %s\n", o.color(ColorBold, "Network Quality:"))
	fmt.Printf("    Video Streaming:  %s\n", o.gradeColor(r.Quality.VideoStreaming))
	fmt.Printf("    Online Gaming:    %s\n", o.gradeColor(r.Quality.Gaming))
	fmt.Printf("    Video Chatting:   %s\n", o.gradeColor(r.Quality.VideoChatting))

	fmt.Println(strings.Repeat("─", 48))

	// Verbose details
	if o.cfg.Verbose {
		o.verboseDetails(r)
	}
}

func (o *Output) gradeColor(grade string) string {
	switch grade {
	case "Great":
		return o.color(ColorGreen, grade)
	case "Good":
		return o.color(ColorGreen, grade)
	case "Okay", "Incomplete":
		return o.color(ColorYellow, grade)
	case "Poor":
		return o.color(ColorRed, grade)
	default:
		return grade
	}
}

func (o *Output) confidenceColor(level, text string) string {
	switch strings.ToLower(level) {
	case "high":
		return o.color(ColorGreen, text)
	case "medium":
		return o.color(ColorYellow, text)
	case "low":
		return o.color(ColorRed, text)
	default:
		return text
	}
}

func formatDirectionalLoss(loss *float64, received, sent int) string {
	if loss == nil || sent <= 0 {
		return " N/A"
	}
	return fmt.Sprintf(" %.2f%% (%d/%d)", *loss, received, sent)
}

func (o *Output) verboseDetails(r *client.Results) {
	fmt.Println()
	fmt.Println(o.color(ColorBold, "LATENCY BREAKDOWN"))
	fmt.Println(strings.Repeat("─", 48))

	// Group latency samples by phase
	unloaded := make([]float64, 0)
	for _, s := range r.LatencySamples {
		if s.Phase == "unloaded" {
			unloaded = append(unloaded, float64(s.RTT.Microseconds())/1000)
		}
	}

	if len(unloaded) > 0 {
		min, max, _ := stats(unloaded)
		med := median(unloaded)
		fmt.Printf("  Unloaded:   %.1f ms (min: %.1f, max: %.1f, median: %.1f)\n",
			r.Summary.LatencyUnloadedMs, min, max, med)
	}

	fmt.Println()
	fmt.Println(o.color(ColorBold, "DOWNLOAD TESTS"))
	fmt.Println(strings.Repeat("─", 48))

	// Group by profile
	dlByProfile := make(map[string][]float64)
	for _, s := range r.ThroughputSamples {
		if s.Direction == "download" {
			dlByProfile[s.Profile] = append(dlByProfile[s.Profile], s.Mbps)
		}
	}

	for profile, speeds := range dlByProfile {
		min, max, avg := stats(speeds)
		fmt.Printf("  %-8s x%d:  avg %.1f Mbps  (min: %.0f, max: %.0f)\n",
			profile, len(speeds), avg, min, max)
	}

	fmt.Println()
	fmt.Println(o.color(ColorBold, "UPLOAD TESTS"))
	fmt.Println(strings.Repeat("─", 48))

	ulByProfile := make(map[string][]float64)
	for _, s := range r.ThroughputSamples {
		if s.Direction == "upload" {
			ulByProfile[s.Profile] = append(ulByProfile[s.Profile], s.Mbps)
		}
	}

	for profile, speeds := range ulByProfile {
		min, max, avg := stats(speeds)
		fmt.Printf("  %-8s x%d:  avg %.1f Mbps  (min: %.0f, max: %.0f)\n",
			profile, len(speeds), avg, min, max)
	}

	if r.PacketLoss != nil && !r.PacketLoss.Unavailable {
		fmt.Println()
		fmt.Println(o.color(ColorBold, "PACKET LOSS TEST"))
		fmt.Println(strings.Repeat("─", 48))
		fmt.Printf("  Frame:      %d bytes, exact binary v1\n", r.PacketLoss.FrameSizeBytes)
		fmt.Printf("  Transaction:%s\n", formatDirectionalLoss(
			&r.PacketLoss.TransactionLossPercent,
			r.PacketLoss.Received,
			r.PacketLoss.Sent,
		))
		fmt.Printf("  Forward:    %s\n", formatDirectionalLoss(
			r.PacketLoss.ForwardLossPercent,
			r.PacketLoss.ForwardReceived,
			r.PacketLoss.ForwardSent,
		))
		fmt.Printf("  Reverse ACK:%s\n", formatDirectionalLoss(
			r.PacketLoss.ReverseAcknowledgementLossPercent,
			r.PacketLoss.AcknowledgementsReceived,
			r.PacketLoss.AcknowledgementsSent,
		))
		fmt.Printf("  Server:     duplicates %d, invalid %d, ACK send failures %d\n",
			r.PacketLoss.DuplicateFrames, r.PacketLoss.InvalidFrames, r.PacketLoss.AckSendFailures)
		fmt.Printf("  RTT:        min %.1f ms, median %.1f ms, p90 %.1f ms\n",
			r.PacketLoss.RTTStatsMs.Min, r.PacketLoss.RTTStatsMs.Median, r.PacketLoss.RTTStatsMs.P90)
		fmt.Printf("  Jitter:     %.1f ms\n", r.PacketLoss.JitterMs)
	}

	fmt.Println()
	fmt.Println(o.color(ColorBold, "MEASUREMENT CONFIDENCE"))
	fmt.Println(strings.Repeat("─", 48))
	fmt.Printf("  Overall:    %s (%d/100)\n", strings.ToUpper(r.TestConfidence.Overall), r.TestConfidence.OverallScore)
	fmt.Printf("  Samples:    windows d=%d u=%d; latency unloaded=%d loaded d=%d u=%d\n",
		r.TestConfidence.Metrics.SampleCount.DownloadWindows,
		r.TestConfidence.Metrics.SampleCount.UploadWindows,
		r.TestConfidence.Metrics.SampleCount.UnloadedLatency,
		r.TestConfidence.Metrics.SampleCount.DownloadLoadedLatency,
		r.TestConfidence.Metrics.SampleCount.UploadLoadedLatency)
	fmt.Printf("  Variability: download %.1f%%, upload %.1f%%, latency %.1f%%\n",
		r.TestConfidence.Metrics.Variability.Download,
		r.TestConfidence.Metrics.Variability.Upload,
		r.TestConfidence.Metrics.Variability.Latency)
	for _, warning := range r.TestConfidence.Warnings {
		fmt.Printf("  Warning:    %s\n", warning)
	}
}

func stats(values []float64) (min, max, avg float64) {
	if len(values) == 0 {
		return 0, 0, 0
	}

	min = values[0]
	max = values[0]
	sum := 0.0

	for _, v := range values {
		if v < min {
			min = v
		}
		if v > max {
			max = v
		}
		sum += v
	}

	avg = sum / float64(len(values))
	return
}

func median(values []float64) float64 {
	if len(values) == 0 {
		return 0
	}

	sorted := make([]float64, len(values))
	copy(sorted, values)
	sort.Float64s(sorted)

	n := len(sorted)
	if n%2 == 0 {
		return (sorted[n/2-1] + sorted[n/2]) / 2
	}
	return sorted[n/2]
}

func optionalFloat(value *float64, format string) string {
	if value == nil {
		return "n/a"
	}
	return fmt.Sprintf(format, *value)
}

// Quiet prints minimal output.
func (o *Output) Quiet(r *client.Results) {
	// Format: download_mbps upload_mbps latency_ms loss_percent
	fmt.Printf("%.1f  %.1f  %.1f  %s\n",
		r.Summary.DownloadMbps,
		r.Summary.UploadMbps,
		r.Summary.LatencyUnloadedMs,
		optionalFloat(r.Summary.PacketLossPercent, "%.2f"))
}

// CSV prints CSV output.
func (o *Output) CSV(r *client.Results) {
	fmt.Println("timestamp,server,download_mbps,upload_mbps,latency_ms,jitter_ms,packet_loss_pct")

	hostname := ""
	if r.Meta != nil {
		hostname = r.Meta.Hostname
	}
	loss := ""
	if r.Summary.PacketLossPercent != nil {
		loss = fmt.Sprintf("%.2f", *r.Summary.PacketLossPercent)
	}
	fmt.Printf("%s,%s,%.1f,%.1f,%.1f,%.1f,%s\n",
		r.Timestamp.UTC().Format(time.RFC3339),
		hostname,
		r.Summary.DownloadMbps,
		r.Summary.UploadMbps,
		r.Summary.LatencyUnloadedMs,
		r.Summary.JitterMs,
		loss)
}

// Spinner provides ASCII spinner animation.
type Spinner struct {
	frames   []string
	current  int
	prefix   string
	stop     chan struct{}
	stopped  bool
	mu       sync.Mutex
	interval time.Duration
}

// NewSpinner creates a new spinner.
func NewSpinner() *Spinner {
	return &Spinner{
		frames:   []string{"|", "/", "-", "\\"},
		interval: 100 * time.Millisecond,
		stop:     make(chan struct{}),
	}
}

// Start starts the spinner with the given prefix.
func (s *Spinner) Start(prefix string) {
	s.mu.Lock()
	s.prefix = prefix
	s.stopped = false
	s.mu.Unlock()

	go func() {
		for {
			select {
			case <-s.stop:
				return
			default:
				s.mu.Lock()
				if !s.stopped {
					fmt.Printf("\r%s %s", s.prefix, s.frames[s.current])
					s.current = (s.current + 1) % len(s.frames)
				}
				s.mu.Unlock()
				time.Sleep(s.interval)
			}
		}
	}()
}

// Stop stops the spinner and prints the final text.
func (s *Spinner) Stop(finalText string) {
	s.mu.Lock()
	s.stopped = true
	s.mu.Unlock()

	select {
	case s.stop <- struct{}{}:
	default:
	}

	fmt.Printf("\r%s\n", finalText)
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: netspeed.c/src/output.c
#include "output.h"

#include "json.h"
#include "stats.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#define RESET "\033[0m"
#define RED "\033[31m"
#define GREEN "\033[32m"
#define YELLOW "\033[33m"
#define BLUE "\033[34m"
#define CYAN "\033[36m"
#define BOLD "\033[1m"

bool output_is_tty(void)
{
    const char *term = getenv("TERM");
    return isatty(STDOUT_FILENO) && !getenv("CI") && (!term || strcmp(term, "dumb") != 0);
}

void output_init(output_t *output, const config_t *config)
{
    memset(output, 0, sizeof(*output));
    output->json_mode = config->json_output;
    output->csv_mode = config->csv_output;
    output->quiet = config->quiet;
    output->verbose = config->verbose;
    output->interactive = output_is_tty() && !output->json_mode && !output->csv_mode && !output->quiet;
    output->use_color = output->interactive && !config->no_color;
}

static const char *color(const output_t *output, const char *code)
{
    return output->use_color ? code : "";
}

static const char *reset(const output_t *output)
{
    return output->use_color ? RESET : "";
}

void output_header(const output_t *output, const char *server_url)
{
    if (output->json_mode || output->csv_mode || output->quiet) return;
    printf("%snetspeed%s %s\n", color(output, BOLD), reset(output), NETSPEED_VERSION);
    printf("Server: %s%s%s\n", color(output, CYAN), server_url, reset(output));
    printf("────────────────────────────────────────────────\n");
}

void output_progress(const output_t *output, const char *stage,
                     int current, int total, double value)
{
    if (output->json_mode || output->csv_mode || output->quiet || total <= 0) return;
    if (!output->interactive) {
        if (output->verbose) printf("%s: %d/%d (%.1f)\n", stage, current, total, value);
        return;
    }
    double fraction = (double)current / (double)total;
    if (fraction < 0) fraction = 0;
    if (fraction > 1) fraction = 1;
    int filled = (int)(fraction * 20);
    char bar[24];
    int position = 0;
    for (int index = 0; index < filled && position < 20; index++) bar[position++] = '=';
    if (position < 20) bar[position++] = '>';
    while (position < 20) bar[position++] = ' ';
    bar[position] = '\0';
    const char *unit = strstr(stage, "latency") ? " ms" :
                       (!strcmp(stage, "download") || !strcmp(stage, "upload")) ? " Mbps" : "";
    printf("\r%-15s [%s] %3.0f%% %.1f%s\033[K", stage, bar, fraction * 100, value, unit);
    fflush(stdout);
}

void output_clear_progress(const output_t *output)
{
    if (output->interactive) {
        printf("\r%70s\r", "");
        fflush(stdout);
    }
}

static const char *grade_color(const output_t *output, const char *grade)
{
    if (!output->use_color) return "";
    if (!strcmp(grade, "Great") || !strcmp(grade, "Good")) return GREEN;
    if (!strcmp(grade, "Okay") || !strcmp(grade, "Incomplete")) return YELLOW;
    return RED;
}

static void print_grade(const output_t *output, const char *label, const char *grade)
{
    printf("    %-18s %s%s%s\n", label, grade_color(output, grade), grade, reset(output));
}

static void print_directional(const char *label, bool available, double loss,
                              int received, int sent)
{
    if (!available || sent <= 0) {
        printf("    %-12s N/A\n", label);
    } else {
        printf("    %-12s %.2f%% (%d/%d)\n", label, loss, received, sent);
    }
}

void output_results(const output_t *output, const results_t *results)
{
    output_clear_progress(output);
    printf("\n────────────────────────────────────────────────\n");
    printf("%s                    RESULTS%s\n", color(output, BOLD), reset(output));
    printf("────────────────────────────────────────────────\n");
    printf("  Download:     %s%.1f Mbps%s\n", color(output, CYAN),
           results->summary.download_mbps, reset(output));
    printf("  Upload:       %s%.1f Mbps%s\n", color(output, CYAN),
           results->summary.upload_mbps, reset(output));
    printf("  Latency:      %s%.1f ms (jitter: %.1f ms)%s\n", color(output, BLUE),
           results->summary.latency_unloaded_ms, results->summary.jitter_ms, reset(output));
    if (!results->packet_loss_present) {
        printf("  Packet Loss:  %sN/A (skipped)%s\n", color(output, YELLOW), reset(output));
    } else if (results->packet_loss.unavailable) {
        printf("  Packet Loss:  %sN/A (%s)%s\n", color(output, YELLOW),
               results->packet_loss.reason, reset(output));
    } else {
        printf("  Packet Loss:  %.2f%% transaction (%d/%d)\n",
               results->packet_loss.transaction_loss_percent,
               results->packet_loss.received, results->packet_loss.sent);
        print_directional("Forward:", results->packet_loss.forward_loss_available,
                          results->packet_loss.forward_loss_percent,
                          results->packet_loss.forward_received,
                          results->packet_loss.forward_sent);
        print_directional("Reverse ACK:", results->packet_loss.reverse_loss_available,
                          results->packet_loss.reverse_acknowledgement_loss_percent,
                          results->packet_loss.acknowledgements_received,
                          results->packet_loss.acknowledgements_sent);
    }
    printf("  Confidence:   %s (%d/100)\n",
           results->test_confidence.overall, results->test_confidence.overall_score);
    printf("────────────────────────────────────────────────\n");
    printf("  %sNetwork Quality:%s\n", color(output, BOLD), reset(output));
    print_grade(output, "Video Streaming:", results->quality.video_streaming);
    print_grade(output, "Online Gaming:", results->quality.gaming);
    print_grade(output, "Video Chatting:", results->quality.video_chatting);
    printf("────────────────────────────────────────────────\n");
}

void output_json(const results_t *results)
{
    char *encoded = results_to_json(results);
    if (!encoded) {
        fprintf(stderr, "{\"error\":\"failed to encode results\"}\n");
        return;
    }
    printf("%s\n", encoded);
    free(encoded);
}

void output_csv(const results_t *results)
{
    printf("timestamp,server,download_mbps,upload_mbps,latency_ms,jitter_ms,packet_loss_pct\n");
    printf("%s,%s,%.1f,%.1f,%.1f,%.1f,",
           results->start_time_rfc3339, results->meta.hostname,
           results->summary.download_mbps, results->summary.upload_mbps,
           results->summary.latency_unloaded_ms, results->summary.jitter_ms);
    if (results->summary.packet_loss_available) printf("%.2f", results->summary.packet_loss_percent);
    printf("\n");
}

void output_quiet(const results_t *results)
{
    printf("%.1f  %.1f  %.1f  ", results->summary.download_mbps,
           results->summary.upload_mbps, results->summary.latency_unloaded_ms);
    if (results->summary.packet_loss_available) printf("%.2f", results->summary.packet_loss_percent);
    else printf("N/A");
    printf("\n");
}

void output_error(const output_t *output, const char *message)
{
    if (output->json_mode) {
        json_writer_t writer;
        json_writer_init(&writer);
        json_start_object(&writer);
        json_kv_string(&writer, "error", message ? message : "unknown error");
        json_end_object(&writer);
        fprintf(stderr, "%s\n", json_writer_string(&writer));
        json_writer_free(&writer);
        return;
    }
    fprintf(stderr, "%sError:%s %s\n", color(output, RED), reset(output),
            message ? message : "unknown error");
}

static void print_samples(const results_t *results, const char *direction)
{
    for (int index = 0; index < results->throughput_count; index++) {
        const throughput_sample_t *sample = &results->throughput_samples[index];
        if (strcmp(sample->direction, direction) != 0) continue;
        if (!strcmp(sample->sample_kind, "window")) {
            printf("  window %d: %.1f Mbps; %d requests × %lld bytes; concurrency %d\n",
                   sample->window_index + 1, sample->mbps, sample->request_count,
                   (long long)sample->chunk_bytes, sample->concurrency);
        } else {
            printf("  %-8s run %d: %.1f Mbps (%lld bytes in %.1f ms)\n",
                   sample->profile, sample->run_index, sample->mbps,
                   (long long)sample->size_bytes, sample->duration_ms);
        }
    }
}

void output_verbose(const output_t *output, const results_t *results)
{
    if (!output->verbose) return;
    printf("\n%sLATENCY BREAKDOWN%s\n", color(output, BOLD), reset(output));
    printf("────────────────────────────────────────────────\n");
    printf("  unloaded median %.1f ms; loaded p90 download %.1f ms; upload %.1f ms\n",
           results->summary.latency_unloaded_ms, results->summary.latency_download_ms,
           results->summary.latency_upload_ms);
    printf("\n%sDOWNLOAD TESTS%s\n", color(output, BOLD), reset(output));
    printf("────────────────────────────────────────────────\n");
    print_samples(results, "download");
    printf("\n%sUPLOAD TESTS%s\n", color(output, BOLD), reset(output));
    printf("────────────────────────────────────────────────\n");
    print_samples(results, "upload");
    if (results->packet_loss_present && !results->packet_loss.unavailable) {
        printf("\n%sPACKET LOSS TEST%s\n", color(output, BOLD), reset(output));
        printf("────────────────────────────────────────────────\n");
        printf("  Frame:       %d bytes, exact binary v1\n", results->packet_loss.frame_size_bytes);
        printf("  Server:      duplicates %d, invalid %d, ACK send failures %d\n",
               results->packet_loss.duplicate_frames, results->packet_loss.invalid_frames,
               results->packet_loss.ack_send_failures);
        printf("  RTT:         min %.1f, median %.1f, p90 %.1f ms; jitter %.1f ms\n",
               results->packet_loss.rtt_stats_ms.min, results->packet_loss.rtt_stats_ms.median,
               results->packet_loss.rtt_stats_ms.p90, results->packet_loss.jitter_ms);
    }
    printf("\n%sMEASUREMENT CONFIDENCE%s\n", color(output, BOLD), reset(output));
    printf("────────────────────────────────────────────────\n");
    printf("  Overall:     %s (%d/100)\n", results->test_confidence.overall,
           results->test_confidence.overall_score);
    printf("  Samples:     windows d=%d u=%d; latency unloaded=%d loaded d=%d u=%d\n",
           results->test_confidence.sample_count.download_windows,
           results->test_confidence.sample_count.upload_windows,
           results->test_confidence.sample_count.unloaded_latency,
           results->test_confidence.sample_count.download_loaded_latency,
           results->test_confidence.sample_count.upload_loaded_latency);
    printf("  Variability: download %.1f%%, upload %.1f%%, latency %.1f%%\n",
           results->test_confidence.variability.download,
           results->test_confidence.variability.upload,
           results->test_confidence.variability.latency);
    for (int index = 0; index < results->test_confidence.warning_count; index++) {
        printf("  Warning:     %s\n", results->test_confidence.warnings[index]);
    }
}
